### PR TITLE
Issue/14

### LINKED ISSUE #14: 4.2.1 P+ Individual Extra Coverage (Binxin): Point 1 - Coverage unit tests for peekNumber() & peekKeyword().
CNN 38, Branch Coverage: 92%, **Branch Coverage after: 97%** (with Jacoco Bug not reporting exceptions)- JsonReader::peekNumber@640-741@.\java\com\google\gson\stream\JsonReader.java
CNN 15, Branch Coverage 85%, **Branch Coverage after: 92%**  - JsonReader::peekKeyword@596-638@.\java\com\google\gson\stream\JsonReader.java
 (Important Note: Jacococ does not report custom throws which leads to the tool reporting lower branch coverage then actual,
this can be tested with custom print statements in the branches. This is a known bug for a long time, see: https://github.com/jacoco/eclemma/issues/61)

**Point 1** requires that each group member writes at least four new or enhanced unit tests.
Since all our five chosen functions have  a coverage of 100% except one function where the only not covered branch is untestable.
I've chosen to implement tests for the peekNumber() and peekKeyword() functions in JsonReader.java.

### PR DESCRIPTION
**Pull request by binsu-kth**
4.2.1 P+ Individual Extra Coverage (Binxin): Point 1 - Coverage unit tests for peekNumber() & peekKeyword(), fixes #14.

CNN 38, Branch Coverage: 92%, Branch Coverage after: 97% (with Jacoco Bug not reporting exceptions)- JsonReader::peekNumber@640-741@.\java\com\google\gson\stream\JsonReader.java
CNN 15, Branch Coverage 85%, Branch Coverage after: 92% - JsonReader::peekKeyword@596-638@.\java\com\google\gson\stream\JsonReader.java
(Important Note: Jacococ does not report custom throws which leads to the tool reporting lower branch coverage then actual,
this can be tested with custom print statements in the branches. This is a known bug for a long time, see: jacoco/eclemma#61)

Point 1 requires that each group member writes at least four new or enhanced unit tests.
Since all our five chosen functions have a coverage of 100% except one function where the only not covered branch is untestable.
I've chosen to implement tests for the peekNumber() and peekKeyword() functions in JsonReader.java.

### CODE REVIEW

File: gson/src/test/java/com/google/gson/JsonParserTest.java [hunk 1]
```diff
@@ -49,7 +48,7 @@ public void testParseUnquotedStringArrayFails() {
   }
 
   public void testParseString() {
-    String json = "{a:10,b:'c'}";
+    String json = "{a:10,b:'c'}";;
```
Comment (kittytinythai) [inline gson/src/test/java/com/google/gson/JsonParserTest.java:51]: Added an extra `;` by mistake?

File: gson/src/test/java/com/google/gson/JsonParserTest.java [hunk 2]
```diff
@@ -116,4 +115,96 @@ public void testReadWriteTwoObjects() throws Exception {
...
+   * If all these 3 conditions holds true, the test will fall to the return statement of PEEKED_NONE.
+   * When PEEKED_NONE is returned, the method will look at the next index of the stack.
+   * The JsonElement.getString() should return correctly parsed string.
+   * Expected, JsonElement == JsonPrimitive and getAsString() = String json.
+   */
+  public void testBranchCoverageLine679() {
+    String json = "1-3";
+    JsonElement e = JsonParser.parseString(json);
+    assertTrue(e.isJsonPrimitive());
+    assertEquals(json,e.getAsString());;
```
Comment (kittytinythai) [inline gson/src/test/java/com/google/gson/JsonParserTest.java]: Added an extra `;` by mistake?

File: gson/src/test/java/com/google/gson/JsonParserTest.java [hunk 3]
```diff
@@ -116,4 +115,96 @@ public void testReadWriteTwoObjects() throws Exception {
...
+   * If all these 2 conditions holds true, the test will fall to the return statement of PEEKED_NONE.
+   * When PEEKED_NONE is returned, the method will look at the next index of the stack.
+   * The JsonElement.getString() should return correctly parsed string.
+   * Expected, JsonElement == JsonPrimitive and getAsString() = String json.
+   */
+  public void testBranchCoverageLine686() {
+      String json = "1+3";
+      JsonElement e = JsonParser.parseString(json);
+      assertTrue(e.isJsonPrimitive());
+      assertEquals(json,e.getAsString());;
```
Comment (kittytinythai) [inline gson/src/test/java/com/google/gson/JsonParserTest.java]: Added an extra `;` by mistake?

File: gson/src/test/java/com/google/gson/JsonParserTest.java [hunk 4]
```diff
@@ -116,4 +115,96 @@ public void testReadWriteTwoObjects() throws Exception {
...
+   * When PEEKED_NONE is returned, the method will look at the next index of the stack.
+   * Since the String has a leading 0, it will continue parsing since it could be a octal sequence.
+   * The parse omits the leading 0 when converting to int
+   * Expected, JsonElement == JsonPrimitive and getAsInt() = 23
+   */
+  public void testBranchCoverageLine715() {
+    String json = "023";
+    JsonElement e = JsonParser.parseString(json);
+    assertTrue(e.isJsonPrimitive());
+    assertEquals(23,e.getAsInt());;
```
Comment (kittytinythai) [inline gson/src/test/java/com/google/gson/JsonParserTest.java]: Added an extra `;` by mistake?

File: gson/src/test/java/com/google/gson/JsonParserTest.java [hunk 5]
```diff
@@ -116,4 +115,96 @@ public void testReadWriteTwoObjects() throws Exception {
...
+   * and if the pos + i:th positoin of the keyword index is less then the limit for all i:s.
+   * The function will then return PEEKED_NONE, since the keyword does not correspond to the letter.
+   * The sequence then continues until the end of the string. Since it is not a keyword, it will return the string "f".
+   * Expected, JsonElement == JsonPrimitive and String json == e.getAsString()).
+   */
+  public void testBranchCoverageLine622() {
+    String json = "f";
+    JsonElement e = JsonParser.parseString(json);
+    assertTrue(e.isJsonPrimitive());
+    assertEquals(json,e.getAsString());;
```
Comment (kittytinythai) [inline gson/src/test/java/com/google/gson/JsonParserTest.java]: Added an extra `;` by mistake?

File: gson/src/test/java/com/google/gson/JsonParserTest.java (hunk 1 above)
Comment (binsu-kth) [reply gson/src/test/java/com/google/gson/JsonParserTest.java:51]: fixed all semicolons.

File: gson/src/test/java/com/google/gson/JsonParserTest.java (hunk 2 above)
Comment (binsu-kth) [reply gson/src/test/java/com/google/gson/JsonParserTest.java]: fixed

File: gson/src/test/java/com/google/gson/JsonParserTest.java (hunk 3 above)
Comment (binsu-kth) [reply gson/src/test/java/com/google/gson/JsonParserTest.java]: fixed

File: gson/src/test/java/com/google/gson/JsonParserTest.java (hunk 5 above)
Comment (binsu-kth) [reply gson/src/test/java/com/google/gson/JsonParserTest.java]: fixed

File: gson/src/test/java/com/google/gson/JsonParserTest.java (hunk 4 above)
Comment (binsu-kth) [reply gson/src/test/java/com/google/gson/JsonParserTest.java]: fixed

**Pull request merged**